File: lib/src/main/java/platinum/ecs/Component.java
package platinum.ecs;

import java.lang.reflect.ParameterizedType;

public abstract class Component<T extends System> {
    protected Entity entity;
    @SuppressWarnings({"unchecked"})
    Class<T> systemType = (Class<T>) ((ParameterizedType)this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
    public boolean canUse(Class<? extends System> tS) {
        return tS == systemType;
    }
    public abstract void init(System system);
    public abstract void update(System system);
    protected <C extends Component<?>> C getComponent(Class<C> type) {
        return entity.getComponent(type);
    }
    protected <C extends Component<?>> boolean hasComponent(Class<C> type) {
        return entity.hasComponent(type);
    }
}

File: lib/src/main/java/platinum/ecs/Component.kt
package platinum.ecs

import java.lang.reflect.ParameterizedType

abstract class Component<T : System?> {
    var entity: Entity? = null
    var systemType = (this.javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<*>
    fun canUse(tS: Class<out System?>): Boolean {
        return tS == systemType
    }

    abstract fun init(system: System?)
    abstract fun update(system: System)
    protected inline fun <reified C : Component<*>> getComponent(): C? {
        return entity!!.getComponent()
    }

    protected inline fun <reified C : Component<*>> hasComponent(): Boolean {
        return entity!!.hasComponent<C>()
    }
}